### ISSUE TITLE
sfdisk: remove mbr, partitions and create empty disk (full reset disk)

### ISSUE THREAD

**Issue by ivanmara555** [util-linux/util-linux#3191]
Sometimes it becomes necessary to completely reset the disk in scripts, for example, to switch to a different partitioning system or   mbr/gpt data damaged or bad.

To do this, the following must be erased: mbr (bootstrap code area, partition table, boot signature), gpt (protective mbr, gpt data structures).

For this purpuses i use (for exampe):
for bios mbr: `sgdisk -Z /dev/vda`  or  `dd if=/dev/zero of=/dev/vda bs=512 count=1`
for uefi gpt:   `sgdisk -Z /dev/vda`

sfdisk provides `--delete` option (like: `sfdisk --delete /dev/vda`) which delete only partitons but leaves mbr, protective mbr (verified with: `dd if=/dev/vda bs=512 count=1 | hexdump -C`).

The proposal is add to sfdisk option like `-Z, --zap-all` which will be fully reset disk. In fact, I propose to repeat the functionality of the `sgdisk -Z /dev/vda`, a detailed description of which is here [man sgdisk](https://man.archlinux.org/man/sgdisk.8#Z,)

I can use sgdisk but prefer sfdisk because his powerfull text scripting.

**Comment by marcosfrm** [comment]
I agree with a new option that functions identically to `--wipe=always`, but also zeros all data from LBA 0 to the first partition sector to eliminate any residual MBR bootloader information. A potential drawback, however, is if the initial partition's starting point deviates significantly from the standard 1 MiB gap. This could lead to an excessively time-consuming zeroing process.

**Comment by marcosfrm** [comment]
`sgdisk -Z` appears to:

- zero-fill the first 34 sectors (or 6 sectors on 4Kn): PMBR, GPT header, GPT part entries
- zero-fill the last 33 sectors (or 5 sectors on 4Kn): backup GPT part entries, backup GPT header

My idea of zeroing out all sectors before the first partition is impractical. We don't know where the first partition might start, so we could end up wiping a huge area. On the other hand, the behavior of `sgdisk -Z` is predictable.

**Comment by karelzak** [comment]
You can use `wipefs -a`; it erases everything, including partition tables.

**Comment by ivanmara555** [comment]
> You can use `wipefs -a`; it erases everything, including partition tables.

No. `wipefs -a` erase only partitions but leaves MBR, PMBR. The utility is not suitable for resetting the disk.

**Comment by ivanmara555** [comment]
> `sgdisk -Z` appears to:
> 
> * zero-fill the first 34 sectors (or 6 sectors on 4Kn): PMBR, GPT header, GPT part entries
> * zero-fill the last 33 sectors (or 5 sectors on 4Kn): backup GPT part entries, backup GPT header

As far as I understood from wiki and the  [UEFI GPT Specification](https://uefi.org/specs/UEFI/2.10/05_GUID_Partition_Table_Format.html) it is necessary and sufficient to zero-fill the first 34 sectors and last 33 sectors (taking into account the disk sector size eg 512 or 4K bites) for any disk including empty for guaranteed reset.

**Comment by marcosfrm** [comment]
Only the protective MBR and the GPT header need to occupy exactly 1 sector each (with the end padded with zeros if necessary). The partition entries are contiguous 16 KiB (128 entries, each occupying 128 bytes), which, on 4Kn disks, is equivalent to 4 sectors. This is what I see implemented in the field.

**Comment by karelzak** [comment]
`wipefs -a` removes magic strings from the partition table to make it incomplete and invisible for standard tools and Linux kernel. For example:
```
fdisk --list /dev/sdc
Disk /dev/sdc: 100 MiB, 104857600 bytes, 204800 sectors
Disk model: scsi_debug      
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 524288 bytes
Disklabel type: gpt
Disk identifier: BA847230-42E5-40BF-847E-EE751BFF2C21

Device     Start    End Sectors Size Type
/dev/sdc1   2048 202751  200704  98M Linux filesystem
```
```
# wipefs -a /dev/sdc
/dev/sdc: 8 bytes were erased at offset 0x00000200 (gpt): 45 46 49 20 50 41 52 54
/dev/sdc: 8 bytes were erased at offset 0x063ffe00 (gpt): 45 46 49 20 50 41 52 54
/dev/sdc: 2 bytes were erased at offset 0x000001fe (PMBR): 55 aa
/dev/sdc: calling ioctl to re-read partition table: Success
```
It removes magic strings from the Primary GPT header, Backup GPT header, and Protective MBR.

**Comment by marcosfrm** [comment]
> > You can use `wipefs -a`; it erases everything, including partition tables.
> 
> No. `wipefs -a` erase only partitions but leaves MBR, PMBR. The utility is not suitable for resetting the disk.

Without the magic bytes, any remaining data in these areas is considered invalid and should be ignored by all implementations. Nevertheless, the `sgdisk -Z` functionality would be useful -- I _think_ it's more or less the same as what `diskpart clean` does in Windows.
